test/acceptance/tests/basic/terraform/basic-install/main.tf

```diff
@@ -266,6 +268,82 @@ module "test_server" {
   consul_client_token_secret_arn = var.secure ? module.acl_controller[0].client_token_secret_arn : ""
   acl_secret_name_prefix         = var.suffix
   consul_ecs_image               = var.consul_ecs_image
+
+  // Test passing both a role resource and role data source objects to make sure both
+  // have the necessary fields ("arn" and "id").
+  task_role                     = aws_iam_role.task
+  execution_role                = data.aws_iam_role.execution
+  additional_task_role_policies = [aws_iam_policy.exec.arn]
+}
+
+// Testing passing task/execution role into mesh-task
+resource "aws_iam_role" "task" {
+  name = "test_server_${var.suffix}_task_role"
+  assume_role_policy = jsonencode({
+    Version = "2012-10-17"
+    Statement = [
+      {
+        Action = "sts:AssumeRole"
+        Effect = "Allow"
+        Principal = {
+          Service = "ecs-tasks.amazonaws.com"
+        }
+      },
+    ]
+  })
+}
+
+
+// Policy to allow `aws execute-command`
+resource "aws_iam_policy" "exec" {
+  name   = "ecs-execute-command-${var.suffix}"
+  path   = "/"
+  policy = <<EOF
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Effect": "Allow",
+      "Action": [
+        "ssmmessages:CreateControlChannel",
+        "ssmmessages:CreateDataChannel",
+        "ssmmessages:OpenControlChannel",
+        "ssmmessages:OpenDataChannel"
+      ],
+      "Resource": [
+        "*"
+      ]
+    }
+  ]
+}
+EOF
+
+  # TODO: don't have permission to add tags
+  # tags = var.tags
+}
+
+resource "aws_iam_role" "execution" {
+  name = "test_server_${var.suffix}_execution_role"
+  path = "/ecs/"
+
+  assume_role_policy = jsonencode({
+    Version = "2012-10-17"
+    Statement = [
+      {
+        Action = "sts:AssumeRole"
+        Effect = "Allow"
+        Principal = {
+          Service = "ecs-tasks.amazonaws.com"
+        }
+      }
+    ]
+  })
+}
+
+// Test using an "existing" role name. Users will add a data source,
+// and pass the data source object into mesh-task.
+data "aws_iam_role" "execution" {
+  name = aws_iam_role.execution.name
 }
```

Comment:
Should we split this kind of test out into its own TF module? My thought is that it might make this basic-install test complicated when we try and excercise all our new behaviours in here.

Reply:
Yeah, I think I can move this bit into a test which runs a Terraform plan to validate this.

----

In general, we need to balance test runtime and test organization... Having quick tests is valuable, and Eric and I were already feeling pain from the current 30m test runtime (mostly from the HCP test).

We've so far favored test runtime... so we've updated the `basic-install` deployment to test a variety of things using only two mesh-tasks and a single Go test, which isn't great. We could improve that by adding additional mesh-tasks to test specific configurations, and by breaking validations up into Go sub-tests. That would help us test more cases with a single deployment while keeping the things somewhat organized.

But, that only goes so far too. Like testing an insecure vs secure config vs HCP servers need to be a separate deployment.

Reply:
So, I only moved the bit where it passes the IAM role data source into a separate test, but I still test passing the iam role resources into mesh-task in basic-install.

I'm realizing now that you probably meant moving all of the "passing roles" test logic into a separate test (i.e. testing passing in a IAM role data source and IAM role resource)?